velox/expression/FunctionSignature.cpp

```diff
@@ -222,14 +215,23 @@ std::string AggregateFunctionSignature::toString() const {
   return out.str();
 }
 
+void AggregateFunctionSignature::validate(
+    std::unordered_set<std::string> usedVariables) {
+  validateBaseTypeAndCollectTypeParams(
+      variables_, intermediateType_, usedVariables, false);
+  FunctionSignature::validate(usedVariables);
+}
+
 FunctionSignaturePtr FunctionSignatureBuilder::build() {
   VELOX_CHECK(returnType_.has_value());
-  return std::make_shared<FunctionSignature>(
+  auto funcSignature = std::make_shared<FunctionSignature>(
       std::move(variables_),
       returnType_.value(),
       std::move(argumentTypes_),
       std::move(constantArguments_),
       variableArity_);
+  funcSignature->validate();
```

Comment:
Validation logic needs to stay in the constructor of the FunctionSignature. FunctionSignatureBuilder is only one way to create FunctionSignature. Is it also possible to use FunctionSignature directly. We need to make sure FunctionSignature is always valid no matter how it was created.

Reply:
@mbasmanova If validation logic is in the constructor, the `FunctionSignature` cannot validate `variables_` because it contains variables used in the `intermediateType_` of `AggregateFunctionSignature`, and the `FunctionSignature` cannot determine whether they are valid.